### PR TITLE
Upkeep multiple performs per block

### PR DESCRIPTION
**Pull request by aodhgan**

### CODE REVIEW

File: contracts/PrizeStrategyUpkeep.sol [hunk 1]
```diff
@@ -70,25 +85,44 @@ contract PrizeStrategyUpkeep is KeeperCompatibleInterface, Ownable {
...
+        require(block.number > _upkeepLastUpkeepBlockNumber + upkeepMinimumBlockInterval, "PrizeStrategyUpkeep::minimum block interval not reached");
+
         address[] memory prizePools = prizePoolRegistry.getAddresses();
 
+      
         uint256 batchCounter = upkeepBatchSize; //counter for batch
+
         uint256 poolIndex = 0;
-        
+        uint8 startAwardCounter = 0;
```
Comment (aodhgan) [inline contracts/PrizeStrategyUpkeep.sol:97]: @asselstine uint8 vs uint256 here?

Comment (asselstine) [reply contracts/PrizeStrategyUpkeep.sol:97]: Given gas costs, a limit of 256 is likely fine!

Then again, it's a local variable so using a full word uint256 won't make a difference in terms of gas

File: contracts/PrizeStrategyUpkeep.sol [hunk 2]
```diff
@@ -70,25 +85,44 @@ contract PrizeStrategyUpkeep is KeeperCompatibleInterface, Ownable {
...
             
             if(prizeStrategy.canStartAward()){
                 PeriodicPrizeStrategyInterface(prizeStrategy).startAward();
+                updatedUpkeepBlockNumber = block.number;
+                startAwardCounter++;
                 batchCounter--;
             }
             else if(prizeStrategy.canCompleteAward()){
                 PeriodicPrizeStrategyInterface(prizeStrategy).completeAward();
+                updatedUpkeepBlockNumber = block.number;
```
Comment (asselstine) [inline contracts/PrizeStrategyUpkeep.sol:114]: I'm thinking you can do this just once outside of the loop by checking completeAwardCounter > 0 or startAwardCounter > 0

File: contracts/PrizeStrategyUpkeep.sol [hunk 3]
```diff
@@ -70,25 +85,44 @@ contract PrizeStrategyUpkeep is KeeperCompatibleInterface, Ownable {
...
+                completeAwardCounter++;
                 batchCounter--;
             }
             poolIndex++;            
         }
+
+        // SSTORE upkeepLastUpkeepBlockNumber once
+        if(_upkeepLastUpkeepBlockNumber != updatedUpkeepBlockNumber){
+            upkeepLastUpkeepBlockNumber = updatedUpkeepBlockNumber;
+            emit UpkeepPerformed(startAwardCounter, startAwardCounter);
```
Comment (asselstine) [inline contracts/PrizeStrategyUpkeep.sol:124]: Looks like one of these should be completeAwardCounter

Comment (asselstine) [reply contracts/PrizeStrategyUpkeep.sol:124]: Would be wise to add an assertion to the test

**Review by asselstine — APPROVED** [review]
Looks good other than a few changes

**Pull request merged**